### ISSUE TITLE
Perhaps we should reduce the default call stack size.

### ISSUE THREAD

**Issue by gbtitus** [chapel-lang/chapel#7492]
The current default call stack size for Chapel tasks is 8mb.  This was originally chosen because it is a common setting for the process stack limit on UNIX/Linux systems and we felt we could avoid surprises by using the same thing for Chapel tasks.

But based on a fair amount of admittedly unstructured experience, very few tasks actually need 8mb of stack space. And we could benefit by reducing the default.  At least in programs without large arrays, stacks are the major driver of total data size because of the way Qthreads, in particular, manages qthread/task stack space by creating pools of stacks per worker thread.

Especially for configurations in which we're limited by a fixed-size heap due to the need to register memory with a network, we have been reducing the call stack size in testing for some time. There are about 25 occurrences of `CHPL_RT_CALL_STACK_SIZE` in `.*execenv` files under the test hierarchy, for example.  And there are cases where network registration doesn't play a role, either. See for example [CHAPEL-316](https://chapel.atlassian.net/browse/CHAPEL-316).

If we do this we'll need to run a much broader than normal range of configurations for testing, especially including many debug modes where there might be additional stack requirements for `printf()` buffers and so on.

**Comment by bradcray** [comment]
I don't have any particular objection to this proposal, but am trying to remember things that caused us to make it so big by default.  I.e., what would users/developers do that would cause stack overflows which were tough for them/us to debug back when we had small callstacks by default for user-level tasking packages like Qthreads?

**Comment by gbtitus** [comment]
The original tasks=fifo default stack size came from `getrlimit(RLIMIT_STACK)`, which returns the process stack size (the common 8mb value).  The original tasks=qthreads default stack size was presumably the Qthreads default (4kb?).  It was changed to `32*1024*sizeof(size_t)` in mid-2011 (commit a8123eec66) to match the nanox tasking layer.  That would have been 256kb on most systems. Then a few years later, in early 2014 (commit 62de84bc17), it was changed to 8mb to "avoid false failures due to stack overflows".  Unfortunately neither commit lists any specific failing tests.  But based on the commit message, the change to 8mb was done so that we could start regular testing with tasks=qthreads.  In my memory of that period the change was a bit of a stopgap and the commit message phrasing seems to support that. There were enough failures with the existing 256kb limit that regular testing would have been annoyingly noisy. But we weren't enthused about doing a hunt for the smallest stack size that would let all the testing succeed, so we decided just to go with what tasks=fifo was using because we knew that was working.

**Comment by bradcray** [comment]
I was thinking less of running within the test harness and more of cases where users or developers would report "I got a terrible failure" and then the result turned out to be stack overflow...   While those weren't an everyday occurrence, it seemed like they were frequent enough that part of the motivation for bumping qthreads up to fifo levels was to make them go away (which they largely have).  In saying that, I'm not suggesting we keep things at 8mb—I was more trying to remember what kinds of use cases led us to these overflows and whether that suggested what the new default should be and/or whether we were less likely to hit such problems with today's Chapel.

**Comment by gbtitus** [comment]
I'm not recalling any such user or developer reports.  But they would have resulted from the 256kb default stacksize tasks=qthreads was using prior to our change to 8mb, so that would definitely indicate 256kb will be too small.  Do you have any recollection of what those tests might have been, if they were things that never made it into our test hierarchy?

Another thing we should keep in mind is that since those days we've moved a fair number of things out of the heap and into the stack, so our stack requirements have actually risen.

I didn't mention it and probably should have, but I'm planning on doing a parametric sweep of stack sizes across the whole test suite as input to proposing any different (smaller) default.  I was envisioning powers of 2 working downward from 8mb.  I'll be surprised if I can get all the way down to the old default of 256kb without a large number of failures.  I'm guessing that's just too small.  I regularly run some of the benchmarks (ra, stream, etc.) with stack=128kb, but I know there are tests in our suites that won't work with anywhere near that little stack space.

**Comment by bradcray** [comment]
Dredging emails, some possible stack hogs within the test suite might be:
* studies/sudoku/dinan/search/sudoku-reduce*.chpl

But I didn't find as much pain and suffering as I thought, so maybe the issue wasn't as big as I'm worried about.  I did find a relevant Jan 2014 mail from Greg with subject "how best to bump stack limits?"

I did find a lot of mail in which we were talking about having stack overflow checking on by default and off when --fast was thrown, which is an idea I still like.  I think we never implemented it, though, did we?  If we had that, I'd be far less concerned about dialing down the stack size for sure...

**Comment by ronawho** [comment]
We have guard pages enabled by default now (unless --fast/--no-stack-checks is thrown or if huge pages are being used.)

That didn't used to be the case, so we didn't get clean error messages. Just segfaults or other non-deterministic faults. We should get reliable error messages with guard pages, although I'm not sure what that error message will look like off the top of my head.

**Comment by bradcray** [comment]
OK, cool.  Bonus points if said error message tells you the name of the env. var. to set to expand the stack and/or points to the appropriate documentation.

**Comment by ronawho** [comment]
Although.. the guard page is just an mprotect'ed page -- so that might still just be a segfault

**Comment by gbtitus** [comment]
Yes, you just get a segfault when you access the inaccessible guard page.  In order to produce an error message we'd have to catch the `SIGSEGV` and then in the signal handler, print the message if the target memory address was in a guard page.  This is certainly doable; among the info in the `siginfo_t` passed for a `SIGSEGV` to a handler installed by `sigaction(2)` is the target address, so you don't even have to decode the failing instruction and compute the address yourself like in the olden days. Yay progress!

**Comment by gbtitus** [comment]
Another commit with stack size implications is 56082bf9 in late 2014.  This changed the qthreads tasking shim to call the bodies of tasks encountered when we were serial directly, rather than enqueueing them to Qthreads.  Doing so saves significant time and space for this case but increases stack size requirements because the child's stack usage gets, well, stacked on top of whatever the parent has already used on its own stack.  I'd found this commit in my earlier search for those applicable to stack requirements, but misplaced my note about it when writing up my initial [comment](https://github.com/chapel-lang/chapel/issues/7492#issuecomment-332943968) above.

**Comment by mppf** [comment]
I just wanted to point out the relationship between this issue and #9984.
